File: libs/EADCSharpClasses/EAD/Cryptography/CSharp/CRC32.cs
namespace EAD.Cryptography.CSharp
{
    using System;
    using System.IO;

    public class CRC32
    {
        private const int BUFFER_SIZE = 0x400;
        private uint[] crc32Table;

        public CRC32()
        {
            uint num = 0xedb88320;
            this.crc32Table = new uint[0x100];
            for (uint i = 0; i < 0x100; i++)
            {
                uint num4 = i;
                for (uint j = 8; j > 0; j--)
                {
                    if ((num4 & 1) == 1)
                    {
                        num4 = (num4 >> 1) ^ num;
                    }
                    else
                    {
                        num4 = num4 >> 1;
                    }
                }
                this.crc32Table[i] = num4;
            }
        }

        public uint GetCrc32(Stream stream)
        {
            uint maxValue = uint.MaxValue;
            byte[] buffer = new byte[0x400];
            int count = 0x400;
            for (int i = stream.Read(buffer, 0, count); i > 0; i = stream.Read(buffer, 0, count))
            {
                for (int j = 0; j < i; j++)
                {
                    maxValue = (maxValue >> 8) ^ this.crc32Table[(int) ((IntPtr) (buffer[j] ^ (maxValue & 0xff)))];
                }
            }
            return ~maxValue;
        }
    }
}

File: libs/EADVBClasses/EAD/Cryptography/VisualBasic/CRC32.vb
Imports System
Imports System.IO

Namespace EAD.Cryptography.VisualBasic
    Public Class CRC32
        ' Methods
        Public Sub New()
            Dim num2 As Integer = -306674912
            Me.crc32Table = New Integer(&H101  - 1) {}
            Dim index As Integer = 0
            Do
                Dim num As Integer = index
                Dim num4 As Integer = 8
                Do
                    If ((num And 1) > 0) Then
                        num = CInt(((CLng((num And -2)) / 2) And &H7FFFFFFF))
                        num = (num Xor num2)
                    Else
                        num = CInt(((CLng((num And -2)) / 2) And &H7FFFFFFF))
                    End If
                    num4 = (num4 + -1)
                Loop While (num4 >= 1)
                Me.crc32Table(index) = num
                index += 1
            Loop While (index <= &HFF)
        End Sub

        Public Function GetCrc32(ByRef stream As Stream) As Integer
            Dim num2 As Integer = -1
            Dim buffer As Byte() = New Byte(&H401  - 1) {}
            Dim count As Integer = &H400
            Dim i As Integer = stream.Read(buffer, 0, count)
            Do While (i > 0)
                Dim num8 As Integer = (i - 1)
                Dim j As Integer = 0
                Do While (j <= num8)
                    Dim index As Integer = ((num2 And &HFF) Xor buffer(j))
                    num2 = (((num2 And -256) / &H100) And &HFFFFFF)
                    num2 = (num2 Xor Me.crc32Table(index))
                    j += 1
                Loop
                i = stream.Read(buffer, 0, count)
            Loop
            Return Not num2
        End Function


        ' Fields
        Private Const BUFFER_SIZE As Integer = &H400
        Private crc32Table As Integer()
    End Class
End Namespace

